### PR TITLE
[MAINTENANCE] Static type checking with python 3.8

### PR DESCRIPTION
**Pull request by anthonyburdi**
Changes proposed in this pull request:
- Run our type checking CI step using python 3.8 in preparation for: pandas 2.0.0 compatibility and python 3.7 EOL 
- Cleans up some typing issues that crop up when running our type checking step with python 3.8
- Adds a `--python-version` option to `invoke type-check` and uses it to explicitly set the python version used for type checking.

### Definition of Done

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

**Comment by ghost** [comment]
<details open>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=7637&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

####

[Review these changes using an interactive CodeSee Map](https://app.codesee.io/r/reviews?pr=7637&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
CodeSee Map legend
</details>

File: ci/azure-pipelines-dev.yml [hunk 1]
```diff
@@ -144,18 +144,23 @@ stages:
 
     jobs:
     - job: static_type_check
+      strategy:
+        matrix:
+          Python38:
+            python.version: '3.8'
```
Comment (cdkini) [inline ci/azure-pipelines-dev.yml:150]: Do we need a matrix if we're just using a single version? Could we just set the `versionSpec` with a hardcoded `3.8` or use Azure variables: https://learn.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch ?

File: great_expectations/compatibility/numpy.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+from typing import Optional, Union
+
+import numpy as np
+from numpy import typing as npt
+from packaging import version
+
+
+def numpy_quantile(
+    a: npt.NDArray, q: float, method: str, axis: Optional[int] = None
+) -> Union[np.float64, npt.NDArray]:
```
Comment (cdkini) [inline great_expectations/compatibility/numpy.py:12]: Nit - `int | None` and `np.float64 | npt.NDArray` (but `ruff` may take care of this?)

File: great_expectations/core/_docs_decorators.py [hunk 3]
```diff
@@ -285,16 +280,17 @@ def _add_text_below_string_docstring_argument(
 
     # Returns: includes an additional ":\n" that we need to strip out.
     if parsed_docstring.returns:
-        parsed_docstring.returns.description = (
-            parsed_docstring.returns.description.strip(":\n")
-        )
+        if parsed_docstring.returns.description:
+            parsed_docstring.returns.description = (
+                parsed_docstring.returns.description.strip(":\n")
```
Comment (cdkini) [inline great_expectations/core/_docs_decorators.py:285]: Should this stripped value be a constant? Is it ever OS dependent or consistent across environments?

File: great_expectations/core/usage_statistics/execution_environment.py [hunk 4]
```diff
@@ -100,11 +101,14 @@ def _get_all_installed_packages(self) -> List[str]:
...
         if not self._all_installed_packages:
             # Only retrieve once
-            self._all_installed_packages = [
-                item.metadata.get("Name").lower()
-                for item in metadata.distributions()
-                if item.metadata.get("Name") is not None
-            ]
+            package_names: list[str] = []
+            for package in metadata.distributions():
+                package_name = package.metadata.get("Name")
```
Comment (cdkini) [inline great_expectations/core/usage_statistics/execution_environment.py:106]: Do we know the schema of `package.metadata`? This is 3rd party so I guess we're okay here

File: ci/azure-pipelines-dev.yml (hunk 1 above)
Comment (anthonyburdi) [reply ci/azure-pipelines-dev.yml:150]: Good catch - originally I had this stage running 3.7 and 3.8. I'll clean it up.

File: ci/azure-pipelines-dev.yml [hunk 5]
```diff
@@ -144,18 +144,23 @@ stages:
...
       steps:
       - task: UsePythonVersion@0
         inputs:
-          versionSpec: 3.7
-        displayName: 'Use Python 3.7'
```
Comment (Shinnnyshinshin) [inline ci/azure-pipelines-dev.yml:151]: ❤️

File: ci/azure-pipelines-dev.yml (hunk 1 above)
Comment (Shinnnyshinshin) [reply ci/azure-pipelines-dev.yml:150]: Would 3.8 or just 3.10 be useful here? Happy to go either way @anthonyburdi and @cdkini

File: great_expectations/cli/util.py [hunk 6]
```diff
@@ -79,13 +79,6 @@ def library_install_load_check(
         f"pip install {pip_library_name}"
     )
 
-    # project_distribution: Distribution = get_project_distribution()
```
Comment (Shinnnyshinshin) [inline great_expectations/cli/util.py:82]: 🚀

File: great_expectations/core/_docs_decorators.py [hunk 7]
```diff
@@ -1,12 +1,7 @@
 from textwrap import dedent
 from typing import Any, Callable, TypeVar
 
-try:
-    import docstring_parser
-    from docstring_parser import DocstringStyle
-except ImportError:
-    docstring_parser = None
-    DocstringStyle = None
+from great_expectations.compatibility import docstring_parser
```
Comment (Shinnnyshinshin) [inline great_expectations/core/_docs_decorators.py:4]: ❤️

File: great_expectations/compatibility/numpy.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    """
+    As of NumPy 1.21.0, the 'interpolation' arg in quantile() has been renamed to `method`.
+    Source: https://numpy.org/doc/stable/reference/generated/numpy.quantile.html
+    """
+    quantile: npt.NDArray
+    if version.parse(np.__version__) >= version.parse("1.22.0"):
```
Comment (Shinnnyshinshin) [inline great_expectations/compatibility/numpy.py:18]: :)

File: great_expectations/util.py [hunk 9]
```diff
@@ -1572,8 +1550,8 @@ def isclose(
     return cast(
         bool,
         np.isclose(
-            a=np.float64(operand_a),  # type:ignore[arg-type]
-            b=np.float64(operand_b),  # type:ignore[arg-type]
+            a=np.float64(operand_a),  # type: ignore[arg-type]
```
Comment (Shinnnyshinshin) [inline great_expectations/util.py:1553]: great catch

**Review by Shinnnyshinshin — APPROVED** [review]
🐼

**Comment by anthonyburdi** [comment]
Whoops apologies for merging too quickly here @cdkini - I'll follow up with a PR to address your comments. (WIP PR https://github.com/great-expectations/great_expectations/pull/7647)

File: great_expectations/compatibility/numpy.py (hunk 2 above)
Comment (anthonyburdi) [reply great_expectations/compatibility/numpy.py:12]: Good catch, fixed in https://github.com/great-expectations/great_expectations/pull/7647. Ruff does not complain, maybe we need to add UP007? https://beta.ruff.rs/docs/rules/#pyupgrade-up

Comment (anthonyburdi) [reply great_expectations/compatibility/numpy.py:12]: Ah looks like we disable that rule: https://github.com/great-expectations/great_expectations/blob/develop/pyproject.toml#L248. We can probably polish and ratchet or mass fix these.

File: ci/azure-pipelines-dev.yml (hunk 1 above)
Comment (anthonyburdi) [reply ci/azure-pipelines-dev.yml:150]: Changed to a variable in https://github.com/great-expectations/great_expectations/pull/7647. 

@Shinnnyshinshin - I think we can go to 3.10 but I'm not sure how big of a lift that will be. I'll check on that separately.

File: great_expectations/core/_docs_decorators.py (hunk 3 above)
Comment (anthonyburdi) [reply great_expectations/core/_docs_decorators.py:285]: This is mainly to clean up the rendered output in our docs build process, so we set up and manage the environment (Netlify). This is a good question, I think if we were expecting to build in different environments we would address. For now I'll leave it as-is.

File: great_expectations/core/usage_statistics/execution_environment.py (hunk 4 above)
Comment (anthonyburdi) [reply great_expectations/core/usage_statistics/execution_environment.py:106]: Yep! See here: https://packaging.python.org/en/latest/specifications/core-metadata/#name.

**Pull request merged**